Index: gcc.fsf.master/gcc/cfgexpand.c
===================================================================
--- gcc.fsf.master.orig/gcc/cfgexpand.c	2009-08-20 14:04:30.060847582 +0200
+++ gcc.fsf.master/gcc/cfgexpand.c	2009-08-20 14:04:36.327849397 +0200
@@ -124,7 +124,7 @@
    GIMPLE tuple STMT and returns the same statement in the form of a
    tree.  */
 
-static tree
+tree
 gimple_to_tree (gimple stmt)
 {
   tree t;
