Index: lto/gcc/cfgexpand.c
===================================================================
--- lto.orig/gcc/cfgexpand.c	2009-08-31 10:39:47.371256259 +0200
+++ lto/gcc/cfgexpand.c	2009-08-31 10:49:52.283196375 +0200
@@ -124,7 +124,7 @@
    GIMPLE tuple STMT and returns the same statement in the form of a
    tree.  */
 
-static tree
+tree
 gimple_to_tree (gimple stmt)
 {
   tree t;
@@ -363,7 +363,7 @@
 
 /* Release back to GC memory allocated by gimple_to_tree.  */
 
-static void
+void
 release_stmt_tree (gimple stmt, tree stmt_tree)
 {
   tree_ann_common_t ann;
